### PR TITLE
nsupdate

### LINKED ISSUE #336: DNS API password with special characters
I've made a custom DNS API script and have a problem saving passwords with special characters, namely `$` and `"`. This problem occurs for any DNS API script!

My simple PR #334 fixes it for these characters, but introduces the problem for `'`.
I know this is edge case, but `$` signs are quite popular in passwords, so I'd opt for the case where `'` isn't allowed.

After some playing around, it would work if the `'` characters were saved as `'"'"'` (single-double-single-double-single quote), or as it would need to be for `_saveaccountconf`: `'\'\"\'\"\''` to escape properly. ([reference](https://stackoverflow.com/a/1250279))

@Neilpang Before this goes too far, what do you think about this?

### PR DESCRIPTION
**Pull request by philfry**
this pull request supersedes pull request #326.
Please close PR #326 as contains obsolete changes.

### CODE REVIEW

File: dnsapi/dns_nsupdate.sh [hunk 1]
```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env bash
```
Comment (Neilpang) [inline dnsapi/dns_nsupdate.sh]: we need to use `#!/usr/bin/env sh`
we must be `sh` compatible,  in freebsd/openbsd etc,  bash is not available by default.

File: dnsapi/dns_nsupdate.sh [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+
+########  Public functions #####################
+
+#Usage: dns_nsupdate_add   _acme-challenge.www.domain.com   "XKrxpRBosdIKFzxW_CT3KLZNf6q0HG9i01zxXp5CPBs"
+dns_nsupdate_add() {
+  fulldomain=$1
+  txtvalue=$2
+  _checkKeyFile || return 1
+  NSUPDATE_SERVER=${NSUPDATE_SERVER:-localhost}
```
Comment (Neilpang) [inline dnsapi/dns_nsupdate.sh]: do not use bash-only grammar :`${NSUPDATE_SERVER:-"localhost"}`
use  `if ... else ..` instead.  we must be `sh` compatible.

File: dnsapi/dns_nsupdate.sh [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  # save the dns server and key to the account conf file.
+  _saveaccountconf NSUPDATE_SERVER "${NSUPDATE_SERVER}"
+  _saveaccountconf NSUPDATE_KEY "${NSUPDATE_KEY}"
+  tmp=$(mktemp --tmpdir acme_nsupdate.XXXXXX)
```
Comment (Neilpang) [inline dnsapi/dns_nsupdate.sh]: Use our internal function `_mktemp()`  instead.
In some platforms, the command `mktemp` is not available.
And the command syntax for `mktemp` is different between GnuLinux and BSD/Mac.

File: dnsapi/dns_nsupdate.sh [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  cat > ${tmp} <<EOF
+server ${NSUPDATE_SERVER}
+update add ${fulldomain}. 60 in txt "${txtvalue}"
+send
+EOF
+  _info "adding ${fulldomain}. 60 in txt \"${txtvalue}\""
+  nsupdate -k ${NSUPDATE_KEY} ${tmp}
+  if [ $? -ne 0 ]; then
```
Comment (Neilpang) [inline dnsapi/dns_nsupdate.sh]: use `if [ "$?" != "0" ] ; ` ,  as space is missing.

File: dnsapi/dns_nsupdate.sh [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  rm -f ${tmp}
+  
+  return 0
+}
+
+#Usage: dns_nsupdate_rm   _acme-challenge.www.domain.com
+dns_nsupdate_rm() {
+  fulldomain=$1
+  _checkKeyFile || return 1
+  NSUPDATE_SERVER=${NSUPDATE_SERVER:-localhost}
```
Comment (Neilpang) [inline dnsapi/dns_nsupdate.sh]: do not use bash-only syntax

File: dnsapi/dns_nsupdate.sh [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  tmp=$(mktemp --tmpdir acme_nsupdate.XXXXXX)
```
Comment (Neilpang) [inline dnsapi/dns_nsupdate.sh]: _mktemp

File: dnsapi/dns_nsupdate.sh [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  cat > ${tmp} <<EOF
+server ${NSUPDATE_SERVER}
+update delete ${fulldomain}. txt
+send
+EOF
+  _info "removing ${fulldomain}. txt"
+  nsupdate -k ${NSUPDATE_KEY} ${tmp}
+  if [ $? -ne 0 ]; then
```
Comment (Neilpang) [inline dnsapi/dns_nsupdate.sh]: space

File: dnsapi/dns_nsupdate.sh [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  rm -f ${tmp}
+
+  return 0
+}
+
+
+####################  Private functions bellow ##################################
+
+_checkKeyFile() {
+  if [ -z "${NSUPDATE_KEY}" ]; then
```
Comment (Neilpang) [inline dnsapi/dns_nsupdate.sh]: `if [ -z "${NSUPDATE_KEY}" ] ;`  space is missing.

File: dnsapi/dns_nsupdate.sh [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    _err "you must specify a path to the nsupdate key file"
+    return 1
+  fi
+  if [ ! -r "${NSUPDATE_KEY}" ]; then
```
Comment (Neilpang) [inline dnsapi/dns_nsupdate.sh]: `if [ ! -r "${NSUPDATE_KEY}" ] ; then` ,  space key missing

File: dnsapi/dns_nsupdate.sh [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    _err "key ${NSUPDATE_KEY} is unreadable"
+    return 1
+  fi
+}
+
+_info() {
```
Comment (Neilpang) [inline dnsapi/dns_nsupdate.sh]: remove those functions.  `_info, _err, _debug, _debug2`

File: dnsapi/dns_nsupdate.sh (hunk 8 above)
Comment (philfry) [inline dnsapi/dns_nsupdate.sh]: that's rather a question of coding style, isn't it? afair the bourne shell (sh) didn't require spaces here.

**Comment by Neilpang** [comment]
Please fix the warnings here:  https://travis-ci.org/Neilpang/acme.sh/builds/175691245

**Comment by Neilpang** [comment]
Can you please update the readme files  to include the nsupdate support ?

https://github.com/philfry/acme.sh/blob/master/dnsapi/README.md

https://github.com/philfry/acme.sh/blob/master/README.md

**Pull request merged**